### DISCUSSION TITLE
Get the field ID of the list item after upload a file and update custom metadata

### DISCUSSION THREAD

**Discussion by developerint3030** [pnp/pnpcore#1141] [Q&A] [answered]
Hi,

I am uploading a file in a document library as follow: 

............
 //Upload file
 var file = await folder.Files.AddAsync(filename, stream, overwrite);

 // Load the file metadata again to get the ListItemFields populated
 await file.LoadAsync(p => p.ListItemAllFields);

 // Update file properties
 file.ListItemAllFields["Title"] = "Test Doc";
 
 file.ListItemAllFields["Field1"] = "example1";
 file.ListItemAllFields["Field2"] = "example2";
 ....

Field1 and Field2 are custom properties for that library and I would like to know what is the best way to check if the custom properties exist in the library. 

Also how I can query the field ID for the list item where the file was uploaded.

Thanks!

**Comment by developerint3030** [comment]
I noticed the ID is available in file.ListItemAllFields.Id

What will be the best way to check if a field is part of the collection fields for that file.

Should I be using IsPropertyAvailable? If so, can I have an example?

Thanks

**Comment by developerint3030** [comment]
I will explain my requirement:

After upload the file, I need to update the metadata, but not all the metadata are available in all the libraries, and I don't know which library I am getting the request to upload the file, so this is what I need to do:

1-Upload the file
2-Load only the fields that exist in the library
3-Update the fields that only exist

and I dont want to load all the fields in the library, just the ones that exist

I hope I make sense

Thanks for your help!

**Reply by jansenbe** [reply]
@developerint3030 : they key task here is to figure out the document library that holds the folder you're uploading the file to. I've copied some snippets from the PnP Core internals:

```csharp

// Get the folder properties, you also load them when you load the folder initially
var tempFolder = await folder.GetAsync(p => p.Properties).ConfigureAwait(false);

if (tempFolder.IsPropertyAvailable(p => p.Properties))
{
    docLibId = DiscoverDocLibId(tempFolder.Properties);
}

private static Guid DiscoverDocLibId(IPropertyValues properties)
{
    string propertyToParse = null;
    if (properties.Values.ContainsKey("vti_listid") && !string.IsNullOrEmpty(properties["vti_listid"].ToString()))
    {
        propertyToParse = properties["vti_listid"].ToString();
    }
    else if (properties.Values.ContainsKey("vti_listname") && !string.IsNullOrEmpty(properties["vti_listname"].ToString()))
    {
        propertyToParse = properties["vti_listname"].ToString();
    }

    if (!string.IsNullOrEmpty(propertyToParse))
    {
        if (Guid.TryParse(propertyToParse, out Guid listGuid))
        {
            return listGuid;
        }
    }

    return Guid.Empty;
}
```

Once you have the document library id you can get a reference to it, including the fields:

```csharp

var myList = await context.Web.Lists.GetByIdAsync(docLibId , p => p.Title,  
                                                     p => p.Fields.QueryProperties(p => p.InternalName, 
                                                                                   p => p.FieldTypeKind, 
                                                                                   p => p.TypeAsString, 
                                                                                   p => p.Title));
``` 

You can now enumerate the fields and use them. To detect whether a field is an OOB field or custom created field is more complex, there's no native solution for that. In another project I've built a list, might be usable for you as well: https://github.com/pnp/pnpassessment/blob/main/src/PnP.Scanning/PnP.Scanning.Core/Scanners/BuiltInFields.cs

**Comment by developerint3030** [comment]
Hi,

Thank you very much for your help.

I did not explain myself very well. I know the name of the library, so I can get the reference to the library:

// Get Document Library List Object
var rootList = (await context.Web.Lists.GetByServerRelativeUrlAsync($"{context.Uri.PathAndQuery}/{libraryName}", p => p.RootFolder));

then I upload the file, and after the upload is completed, I loop through the ListItemAllFields dictionary and if the field exist then I update the value:

// Load the file metadata again to get the ListItemFields populated
await fileAdded.LoadAsync(p => p.ListItemAllFields);

// Update properties
foreach (var property in properties)
{
if (fileAdded.ListItemAllFields.Values.ContainsKey(property.Key))
      {
      		fileAdded.ListItemAllFields[property.Key] = property.Value;
      }
}

But I need to load all the properties to check if the property is in the dictionary or not.

My question is: is there a better way to load just the set of custom properties that exist for that particular library?
Every library can have different set of custom properties.

I get from the client:
1 - Library Name
2 - Custom properties in a key-value collection (in code above the collection is properties)

Thank you so much

**Reply by jansenbe** [reply]
@developerint3030 : I think I've answered your questions above already. Custom properties are fields, so when you load the fields in a library you know all the properties. There's no API that will indicate if a field is something added by SharePoint (an OOB field) or a custom field (added by the user/developer). For that the approach I mentioned is a solution, although not 100% perfect.

**Comment by developerint3030** [comment, marked as answer]
Ok, it makes sense, thanks!
